### ISSUE TITLE
Help is not helpful for error

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#37338]
If you try to compile the code ([play link](https://play.rust-lang.org/?code=fn%20main%28%29%20%7B%0A%20%20%20%20vec!%5B0.0%5D.iter%28%29.map%28%7Cs%7C%20s%20as%20i16%29.collect%3A%3A%3CVec%3Ci16%3E%3E%28%29%3B%0A%7D&version=nightly&backtrace=0)):

``` rust
vec![0.0].iter().map(|s| s as i16).collect::<Vec<i16>>();
```

You get the error

```
rustc 1.14.0-nightly (6e8f92f11 2016-10-07)
error: casting `&f64` as `i16` is invalid
 --> <anon>:2:30
  |
2 |     vec![0.0].iter().map(|s| s as i16).collect();
  |                              ^^^^^^^^
  |
  = help: cast through a raw pointer first

error: aborting due to previous error
```

But casting through a raw pointer (`s as *const f64 as *const i16 as i16`) is not the idiomatic way IMO. Instead it should be fixed by simply dereferencing:

``` rust
vec![0.0].iter().map(|s| *s as i16).collect();
```

**Comment by fhartwig** [comment]
> But casting through a raw pointer (s as *const f64 as *const i16 as i16) is not the idiomatic way IMO. 

It's not just unidiomatic, it actually does something completely different.

**Comment by GuillaumeGomez** [comment]
Funny thing: casting *const f32 as i16 doesn't throw errors. I replace this error with another one which fit better.
